### PR TITLE
Prioritize scheduling build-side of joins

### PR DESCRIPTION
**Pull request by mwd410**
## Description

This modifies the topological scheduling order of SubPlans so that the build side is prioritized for all join types.

The intent here is to optimize query performance to ensure the "build" (i.e. right) side (which must be complete before the join can progress _at all_) is scheduled ahead of the "probe" (i.e. left) side.

## Release notes

(X) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java [hunk 1]
```diff
@@ -1178,7 +1185,63 @@ private StageExecution getStageExecution(StageId stageId)
...
+                        node.getProbeSource().accept(this, context);
+                        return null;
+                    }
+
+                    @Override
+                    public Void visitRemoteSource(RemoteSourceNode node, Void context)
+                    {
+                        for (PlanFragmentId fragmentId : node.getSourceFragmentIds()) {
+                            SubPlan child = sourceSubPlans.remove(fragmentId);
+                            requireNonNull(child, "PlanFragmentId %s does not appear in children of %s".formatted(fragmentId, node));
```
Comment (linzebing) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java:1233]: Nit:
```suggestion
                            requireNonNull(child, "PlanFragmentId %s does not appear in sources of %s".formatted(fragmentId, node));
```

**Review by linzebing — APPROVED** [review]
Do we have unit test coverage for this?

Also, it might be beneficial to do some small benchmarks on this.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java (hunk 1 above)
Comment (mwd410) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java]: oh boy, i can take a look at unit tests, but that sounds difficult.

I completely agree about benchmarks - i would love to learn how to test a local build of trino in a cluster. Maybe you can help me out with that?

**Comment by sopel39** [comment]
@mwd410 thanks!. did you observe perf improve with this change?

**Review by linzebing — APPROVED** [review]

**Review by losipiuk — APPROVED** [review]

**Comment by mwd410** [comment]
Closing as it was merged with #17623

**Pull request closed without merging**